File: 12/gpt-5.1/python/solution.py
def parse_input(path: str):
    with open(path, "r", encoding="utf-8") as f:
        lines = [line.rstrip("\r\n") for line in f]

    shape_sizes: dict[int, int] = {}
    region_lines: list[str] = []
    current_shape_id: int | None = None
    current_shape_count = 0

    for line in lines:
        if not line.strip():
            continue

        if line[0].isdigit() and ":" in line and "x" not in line:
            if current_shape_id is not None:
                shape_sizes[current_shape_id] = current_shape_count
            current_shape_id = int(line.split(":", 1)[0])
            current_shape_count = 0
        elif "x" in line and ":" in line:
            if current_shape_id is not None:
                shape_sizes[current_shape_id] = current_shape_count
                current_shape_id = None
                current_shape_count = 0
            region_lines.append(line.strip())
        elif current_shape_id is not None and ("#" in line or "." in line):
            current_shape_count += line.count("#")

    if current_shape_id is not None:
        shape_sizes[current_shape_id] = current_shape_count

    return shape_sizes, region_lines


def solve_part1(path: str) -> int:
    shape_sizes, region_lines = parse_input(path)

    count_fit = 0

    for region_line in region_lines:
        size_part, counts_part = region_line.split(":", 1)
        size_part = size_part.strip()
        counts_part = counts_part.strip()

        width_str, height_str = size_part.split("x")
        width = int(width_str)
        height = int(height_str)

        counts = [int(x) for x in counts_part.split() if x]

        total_cells_needed = 0
        for shape_id, c in enumerate(counts):
            if c == 0:
                continue
            total_cells_needed += c * shape_sizes[shape_id]

        grid_area = width * height
        if total_cells_needed <= grid_area:
            count_fit += 1

    return count_fit


if __name__ == "__main__":
    part1 = solve_part1("input.txt")
    print(f"Day 12 Part 1: {part1}")

    # Part 2 placeholder
    part2 = 0
    print(f"Day 12 Part 2: {part2}")

File: 12/gpt-5.1/typescript/solution.ts
import * as fs from 'fs';

function parseInput(path: string): { shapeSizes: Map<number, number>; regionLines: string[] } {
	const raw = fs.readFileSync(path, 'utf-8');
	const lines = raw.split(/\r?\n/).map(l => l.replace(/\r$/, ''));

	const shapeSizes = new Map<number, number>();
	const regionLines: string[] = [];

	let currentShapeId = -1;
	let currentShapeCount = 0;

	for (const rawLine of lines) {
		const line = rawLine.trimEnd();
		if (line.trim() === '') continue;

		if (/^\d+:/.test(line) && !line.includes('x')) {
			if (currentShapeId >= 0) {
				shapeSizes.set(currentShapeId, currentShapeCount);
			}
			const idPart = line.split(':', 1)[0];
			currentShapeId = parseInt(idPart, 10);
			currentShapeCount = 0;
		} else if (line.includes('x') && line.includes(':')) {
			if (currentShapeId >= 0) {
				shapeSizes.set(currentShapeId, currentShapeCount);
				currentShapeId = -1;
				currentShapeCount = 0;
			}
			regionLines.push(line.trim());
		} else if (currentShapeId >= 0 && (line.includes('#') || line.includes('.'))) {
			currentShapeCount += (line.match(/#/g) || []).length;
		}
	}

	if (currentShapeId >= 0) {
		shapeSizes.set(currentShapeId, currentShapeCount);
	}

	return { shapeSizes, regionLines };
}

function solvePart1(path: string): number {
	const { shapeSizes, regionLines } = parseInput(path);
	let countFit = 0;

	for (const regionLine of regionLines) {
		const [sizePartRaw, countsPartRaw] = regionLine.split(':', 2);
		const sizePart = sizePartRaw.trim();
		const countsPart = (countsPartRaw ?? '').trim();

		const [wStr, hStr] = sizePart.split('x');
		const width = parseInt(wStr, 10);
		const height = parseInt(hStr, 10);

		const countStrings = countsPart.split(/\s+/).filter(s => s.length > 0);
		let totalCellsNeeded = 0;

		countStrings.forEach((s, idx) => {
			const c = parseInt(s, 10);
			if (!Number.isFinite(c) || c === 0) return;
			const size = shapeSizes.get(idx) ?? 0;
			totalCellsNeeded += c * size;
		});

		const gridArea = width * height;
		if (totalCellsNeeded <= gridArea) {
			countFit++;
		}
	}

	return countFit;
}

const part1 = solvePart1('input.txt');
console.log(`Day 12 Part 1: ${part1}`);

// Part 2 placeholder
const part2 = 0;
console.log(`Day 12 Part 2: ${part2}`);
